File: Data Structures & Algorithms/permutations/submission-5.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res, path = [], []
        used = [False] * len(nums)

        def backtrack(start):
            if start == len(nums):
                res.append(path[:])
                return

            for i in range(len(nums)):
                if used[i]:
                    continue
                path.append(nums[i])
                used[i] = True
                backtrack(start+1)
                path.pop()
                used[i] = False
        backtrack(0)
        return res

File: Data Structures & Algorithms/combination-target-sum/submission-5.java
class Solution {
    public List<List<Integer>> combinationSum(int[] nums, int target) {
        List<List<Integer>> res = new ArrayList<>();
        backtrack(nums, 0, target, new ArrayList<>(), res);
        return res;
    }

    private void backtrack(int[] candidates, int start, int remain, List<Integer> temp, List<List<Integer>> res) {
        if (remain < 0) {
            return;
        }

        if (remain == 0) {
            res.add(new ArrayList<>(temp));
            return;
        }

        for (int i=start; i<candidates.length; i++) {
            temp.add(candidates[i]);
            backtrack(candidates, i, remain - candidates[i], temp, res);
            temp.remove(temp.size()-1);
        }
    }
}
